Kernel/Devices/KCOVDevice.cpp

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2021, Patrick Meyer <git@the-space.agency>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Assertions.h>
+#include <AK/Format.h>
+#include <AK/NonnullOwnPtr.h>
+#include <Kernel/Devices/KCOVDevice.h>
+#include <Kernel/Devices/KCOVInstance.h>
+#include <Kernel/FileSystem/FileDescription.h>
+#include <Kernel/Sections.h>
+#include <Kernel/VM/AnonymousVMObject.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+#include <Kernel/Panic.h>
+
+namespace Kernel {
+
+HashMap<ProcessID, KCOVInstance*>* KCOVDevice::proc_instance;
+HashMap<ThreadID, KCOVInstance*>* KCOVDevice::thread_instance;
+
+UNMAP_AFTER_INIT NonnullRefPtr<KCOVDevice> KCOVDevice::must_create()
+{
+    return adopt_ref_if_nonnull(new KCOVDevice).release_nonnull();
+}
+
+UNMAP_AFTER_INIT KCOVDevice::KCOVDevice()
+    : BlockDevice(30, 0)
+{
+    proc_instance = new HashMap<ProcessID, KCOVInstance*>();
+    thread_instance = new HashMap<ThreadID, KCOVInstance*>();
+    dbgln("KCOVDevice created");
+}
+
+void KCOVDevice::free_thread()
+{
+    auto thread = Thread::current();
+    auto tid = thread->tid();
+
+    auto maybe_kcov_instance = thread_instance->get(tid);
+    if (!maybe_kcov_instance.has_value())
+        return;
+
+    auto kcov_instance = maybe_kcov_instance.value();
+    VERIFY(kcov_instance->state == KCOVInstance::TRACING);
+    kcov_instance->state = KCOVInstance::OPENED;
+    thread_instance->remove(tid);
+}
+
+void KCOVDevice::free_process()
+{
+    auto process = Process::current();
+    auto pid = process->pid();
+
+    auto maybe_kcov_instance = proc_instance->get(pid);
+    if (!maybe_kcov_instance.has_value())
+        return;
+
+    auto kcov_instance = maybe_kcov_instance.value();
+    VERIFY(kcov_instance->state == KCOVInstance::OPENED);
+    kcov_instance->state = KCOVInstance::UNUSED;
+    proc_instance->remove(pid);
+}
+
+KResultOr<NonnullRefPtr<FileDescription>> KCOVDevice::open(int options)
+{
+    auto process = Process::current();
+    if (!process)
+        PANIC("Couldn't get caller process while initializing KCOV instance :^(");
+    auto pid = process->pid();
+    if (proc_instance->get(pid).has_value())
+        return EBUSY; // This process already open()ed the kcov device
+    auto kcov_instance = new KCOVInstance(pid);
```

Comment:
When is this deleted? Seems the KCOVInstance could be RefCounted and stored as a RefPtr in the HashMaps, or it could be OwnPtr'd by the Process map?

Reply:
I imagined deletion to work like this: KCOVInstances are only permanently referenced in the proc_instance and thread_instance HashMaps. When the proc dies, all references will be released and hence the given KCOVInstance is automatically deleted by C++. Let me know if this is nonsense 🤔

Reply:
That would make sense if you were holding them in the hash map with smart pointers that had deleters on them. But since it's only as a set of raw pointers, when the last reference to them goes away they are leaked.

Reply:
A thanks for enlightening me.

At first I thought RefPtr would be the way to go. However I tested around a bit. My kernel reboots with no error message at line `if (kcov_instance->state < KCOVInstance::TRACING) [[likely]]` in SanCop.cpp, while I'm in `thread_instance->remove(tid);` in the KCOV_DISABLE ioctl in KCOVDevice.cpp. I can manually call __sanitizer_cov_trace_pc before and after `thread_instance->remove(tid);` w/o crashing the kernel, as long as I disable the instrumentation. This is likely another case of something recursively calling __sanitizer_cov_trace_pc, more gdb-ing required I guess.

However, I've been thinking about this for a bit. We can't instrument anything that is used by __sanitizer_cov_trace_pc. Therefore we should use as little productively used code as possible in that function and even consider duplicating code to keep the instrumentation for the production build of the kernel where possible. As I only ever want to delete KCOVInstance at process exit, I have now instead opted to manually `delete` the KCOVInstance at the end of KCOVDevice::free_process. It's not as pretty, but still correct.